### PR TITLE
Add sourcegen support for required & init-only properties.

### LINKED ISSUE #58770: System.Text.Json source generator fails to deserialize init-only & required properties
### Description

When using the System.Text.Json source generator, init-only properties are not deserialized.

Repro code:

```cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace JsonSerialization
{
	public static class Program
	{
		public static void Main()
		{
			const string json = @"{ ""InitOnly"": "".NET"", ""Settable"": ""Rocks"" }";

			var tagLine1 = JsonSerializer.Deserialize<TagLine>(json);
			Console.WriteLine(tagLine1.InitOnly ?? "tagLine1.InitOnly is null");
			Console.WriteLine(tagLine1.Settable ?? "tagLine1.Settable is null");

			Console.WriteLine();

			var tagLine2 = JsonSerializer.Deserialize<TagLine>(json, TagLineJsonContext.Default.TagLine);
			Console.WriteLine(tagLine2.InitOnly ?? "tagLine2.InitOnly is null");
			Console.WriteLine(tagLine2.Settable ?? "tagLine2.Settable is null");
		}
	}

	public sealed class TagLine
	{
		public string InitOnly { get; init; }   // This is always null when deserialized with TagLineJsonContext
		public string Settable { get; set; }
	}

	[JsonSerializable(typeof(TagLine), GenerationMode = JsonSourceGenerationMode.Metadata)]
	public partial class TagLineJsonContext : JsonSerializerContext
	{
	}
}
```

Expected output:

```
.NET
Rocks

.NET
Rocks
```

Actual output:

```
.NET
Rocks

tagLine2.InitOnly is null
Rocks
```

### Configuration

* .NET 5.0.9 with System.Text.Json version 6.0.0-preview.7.21377.19
* Windows 10.0.19043.1052 x64

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Updates the source generator to include support for `required` and `init` properties. This is done by piggybacking on the existing parameteric constructor infrastructure using property initializer syntax. For example, given the following type:
```C#
public class MyPoco
{
    public required int X { get; set; }
    public int Y { get; init; }
}
```
The source generator will now generate metadata treating `X` and `Y` as pseudo-constructor parameters and emit a delegate as follows:
```C#
ObjectWithParameterizedConstructorCreator = static (args) => new MyPoco() { X = (int)args[0], Y = (int)args[1] }
```

Fix #58770.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs [hunk 1]
```diff
@@ -51,43 +49,31 @@ internal set
         /// Indicates whether pre-generated serialization logic for types in the context
         /// is compatible with the run time specified <see cref="JsonSerializerOptions"/>.
         /// </summary>
-        internal bool CanUseSerializationLogic
+        internal bool CanUseFastPathSerializationLogic(JsonSerializerOptions options)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:52]: Moves this flag from `JsonSerializerContext` to `JsonSerializerOptions`. This accounts for the fact that, since .NET 7, a `JsonSerializerContext` can be referenced by multiple `JsonSerializerOptions` instances, as such the invalidation logic needs to be evaluated independently for the configuration of every `JsonSerializerOptions` instance.

While not strictly related to this change, I added this in order to simplify the source gen testing infrastructure (removing the requirement for a `options => MyContext(options)` delegate being passed on every test suite declaration.

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs [hunk 2]
```diff
@@ -243,13 +243,13 @@ public async Task TestJsonPathDoesNotFailOnMultiThreads()
             await Task.WhenAll(tasks);
         }
 
-        private async void TestIdTask()
+        private async Task TestIdTask()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.Deserialize.cs:246]: Removes any `async void` incidents per https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 3]
```diff
@@ -0,0 +1,597 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:4]: This is @krwq's original test suite for require member support, moved into and adapted for the `Common` tests folder. Unfortunately github's diff shows this as a new file 😕

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:4]: If you're concerned about git history I believe it is still preserved, so that should be fine.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:52]: nit: if it's not too disruptive can you please undo the test changes that don't concern this feature and make them in a follow up PR?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 4]
```diff
@@ -1173,8 +1189,13 @@ private bool IsValidDataExtensionPropertyType(Type type)
                 }
             }
 
-            private static bool PropertyIsConstructorParameter(PropertyGenerationSpec propSpec, ParameterGenerationSpec[]? paramGenSpecArray)
-                => paramGenSpecArray != null && paramGenSpecArray.Any(paramSpec => propSpec.ClrName.Equals(paramSpec.ParameterInfo.Name, StringComparison.OrdinalIgnoreCase));
+            private static ParameterGenerationSpec? GetMatchingConstructorParameter(PropertyGenerationSpec propSpec, ParameterGenerationSpec[]? paramGenSpecArray)
+            {
+                return paramGenSpecArray?.FirstOrDefault(MatchesConstructorParameter);
+
+                bool MatchesConstructorParameter(ParameterGenerationSpec paramSpec)
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1196]: Were any edge cases considered where this binding logic might not hold true?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 5]
```diff
@@ -959,27 +984,43 @@ private static bool ShouldIncludePropertyForFastPath(PropertyGenerationSpec prop
...
+                if (propertyInitializers != null)
+                {
+                    Debug.Assert(propertyInitializers.Count > 0);
+                    sb.Append("{ ");
+                    foreach (PropertyInitializerGenerationSpec property in propertyInitializers)
+                    {
+                        sb.Append($"{property.Property.ClrName} = {GetParamUnboxing(property.Property.TypeGenerationSpec.Type, property.ParameterIndex)}, ");
+                    }
+
+                    sb.Length -= 2; // delete the last ", " token
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1016]: Love this trick for this classic problem.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:52]: I should have clarified in my earlier comment that this is part of a wider refactoring of the "JsonSerializerWrapper" source gen implementations, which were necessary to accommodate some of the tests in the existing required members test suite.

TL;DR it involves implementing a few missing methods and moving from the JsonSerializerContext overloads to the JsonSerializerOptions. These are functionally equivalent since the removal of reflection fallback in .NET 7. The diff of the JsonSerializerWrapper should clarify what exactly got updated.

So to answer your original question, even though I could remove them, they would necessitate refactoring of the test suite itself.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1196]: It's using the same binding logic as before, the method has been refactored somewhat to also return the matching element.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:52]: No I get that. It just seemed noisy when I first started reading but hey I got through it.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:4]: Makes sense, although it doesn't make reviewing easier :-/

**Comment by eiriktsarpalis** [comment]
The test failures appear to be impacting release builds of wasm only. I'm able to reproduce the failures locally, and all seem related to an issue in which `JsonSerializerOptions.ReferenceHandler` is being ignored. The error appears consistently and only occurs if the entire test suite is being run -- individual test runs are passing without issue. I'll keep investigating but at this point I'm running out of theories.

**Comment by layomia** [comment]
Sounds like the options configuration/instances might not be flowing through properly following the change to move the fast-path decision logic to the context (or just how the context/options are being instantiated). Since individual runs pass my mind also goes to threading issues with the options caching optimizations added in .NET 7. Not sure if these were considered but wanted to mention.

**Comment by eiriktsarpalis** [comment]
> Since individual runs pass my mind also goes to threading issues with the options caching optimizations added in .NET 7. Not sure if these were considered but wanted to mention.

Had the same thought, but the errors keep popping up even with global caching disabled.

Something stranger than that is happening. I've been able to isolate the test failures to using the `Roslyn4.4` target when running tests: https://github.com/dotnet/runtime/pull/79890

Steps to reproduce (on WSL) :
```bash
./build.sh -os Browser -configuration Release && 
./dotnet.sh build -t:Test src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/System.Text.Json.SourceGeneration.Roslyn4.4.Tests.csproj \
    -p:TargetOS=Browser -p:TargetArchitecture=wasm -p:Configuration=Release
```

**Comment by eiriktsarpalis** [comment]
PR blocked until https://github.com/dotnet/runtime/issues/79943 can be addressed.

**Comment by eiriktsarpalis** [comment]
Given that https://github.com/dotnet/runtime/issues/79943 has been fixed, rebasing changes to see if it resolves CI issues.

**Review by layomia — APPROVED** [review]

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.TestLibrary/System.Text.Json.TestLibrary.Roslyn4.4.csproj [hunk 6]
```diff
@@ -3,6 +3,6 @@
   <Import Project="System.Text.Json.TestLibrary.targets" />
 
   <ItemGroup>
-    <ProjectReference Include="..\..\gen\System.Text.Json.SourceGeneration.Roslyn4.0.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\..\gen\System.Text.Json.SourceGeneration.Roslyn4.4.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
```
Comment (ViktorHofer) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.TestLibrary/System.Text.Json.TestLibrary.Roslyn4.4.csproj:6]: Same question here, don't you need the `SetTargetFramework="TargetFramework=netstandard2.0"` attribute? Without it, you are consuming the unshipping .NETCoreApp analyzer. I wonder why this even works today.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/System.Text.Json.SourceGeneration.Roslyn4.4.Unit.Tests.csproj [hunk 7]
```diff
@@ -12,6 +12,6 @@
   <Import Project="System.Text.Json.SourceGeneration.Unit.Tests.targets" />
 
   <ItemGroup>
-    <ProjectReference Include="..\..\gen\System.Text.Json.SourceGeneration.Roslyn4.0.csproj" />
+    <ProjectReference Include="..\..\gen\System.Text.Json.SourceGeneration.Roslyn4.4.csproj" />
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/System.Text.Json.SourceGeneration.Roslyn4.4.Unit.Tests.csproj:15]: It's not. I'll follow up with a new PR amending this.
